File: tools/perf/metrics/smoothness.py
# Copyright 2013 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.
import os

from telemetry.core import util

class SmoothnessMetrics(object):
  def __init__(self, tab):
    self._tab = tab
    with open(
      os.path.join(os.path.dirname(__file__),
                   'smoothness.js')) as f:
      js = f.read()
      tab.ExecuteJavaScript(js)

  def Start(self):
    self._tab.ExecuteJavaScript(
        'window.__renderingStats = new __RenderingStats();'
        'window.__renderingStats.start()')

  def SetNeedsDisplayOnAllLayersAndStart(self):
    self._tab.ExecuteJavaScript(
        'chrome.gpuBenchmarking.setNeedsDisplayOnAllLayers();'
        'window.__renderingStats = new __RenderingStats();'
        'window.__renderingStats.start()')

  def Stop(self):
    self._tab.ExecuteJavaScript('window.__renderingStats.stop()')

  def BindToAction(self, action):
    # Make the scroll test start and stop measurement automatically.
    self._tab.ExecuteJavaScript(
        'window.__renderingStats = new __RenderingStats();')
    action.BindMeasurementJavaScript(self._tab,
                                     'window.__renderingStats.start();',
                                     'window.__renderingStats.stop();')

  @property
  def start_values(self):
    return self._tab.EvaluateJavaScript(
      'window.__renderingStats.getStartValues()')

  @property
  def end_values(self):
    return self._tab.EvaluateJavaScript(
      'window.__renderingStats.getEndValues()')

  @property
  def deltas(self):
    return self._tab.EvaluateJavaScript(
      'window.__renderingStats.getDeltas()')

def Average(numerator, denominator, scale = None, precision = None):
  if denominator == 0:
    return 0
  avg = float(numerator) / float(denominator)
  if scale:
    avg *= scale
  if precision:
    avg = round(avg, precision)
  return avg

def CalcFirstPaintTimeResults(results, tab):
  if tab.browser.is_content_shell:
    results.Add('first_paint', 'ms', 'unsupported')
    return

  tab.ExecuteJavaScript("""
      window.__rafFired = false;
      window.webkitRequestAnimationFrame(function() {
          window.__rafFired  = true;
      });
  """)
  util.WaitFor(lambda: tab.EvaluateJavaScript('window.__rafFired'), 60)

  first_paint_secs = tab.EvaluateJavaScript(
      'window.chrome.loadTimes().firstPaintTime - ' +
      'window.chrome.loadTimes().startLoadTime')

  results.Add('first_paint', 'ms', round(first_paint_secs * 1000, 1))

def CalcResults(benchmark_stats, results):
  s = benchmark_stats

  # Scroll Results
  results.Add('mean_frame_time', 'ms',
              Average(s.total_time, s.screen_frame_count, 1000, 3))
  results.Add('dropped_percent', '%',
              Average(s.dropped_frame_count, s.screen_frame_count,
                      100, 1),
              data_type='unimportant')
  results.Add('percent_impl_scrolled', '%',
              Average(s.impl_thread_scroll_count,
                      s.impl_thread_scroll_count +
                      s.main_thread_scroll_count,
                      100, 1),
              data_type='unimportant')
  results.Add('average_num_layers_drawn', '',
              Average(s.drawn_layers_count, s.screen_frame_count, 1, 1),
              data_type='unimportant')
  results.Add('average_num_missing_tiles', '',
              Average(s.missing_tile_count, s.screen_frame_count, 1, 1),
              data_type='unimportant')

  # Texture Upload Results
  results.Add('average_commit_time', 'ms',
              Average(s.commit_time, s.commit_count, 1000, 3),
              data_type='unimportant')
  results.Add('texture_upload_count', 'count',
              s.texture_upload_count)
  results.Add('total_texture_upload_time', 'seconds',
              s.texture_upload_time)

  # Image Decoding Results
  results.Add('total_deferred_image_decode_count', 'count',
              s.deferred_image_decode_count,
              data_type='unimportant')
  results.Add('total_image_cache_hit_count', 'count',
              s.deferred_image_cache_hits,
              data_type='unimportant')
  results.Add('average_image_gathering_time', 'ms',
              Average(s.image_gathering_time, s.image_gathering_count,
                      1000, 3),
              data_type='unimportant')
  results.Add('total_deferred_image_decoding_time', 'seconds',
              s.deferred_image_decode_time,
              data_type='unimportant')

  # Tile Analysis Results
  results.Add('total_tiles_analyzed', 'count',
              s.tile_analysis_count,
              data_type='unimportant')
  results.Add('solid_color_tiles_analyzed', 'count',
              s.solid_color_tile_analysis_count,
              data_type='unimportant')
  results.Add('average_tile_analysis_time', 'ms',
              Average(s.tile_analysis_time, s.tile_analysis_count,
                      1000, 3),
              data_type='unimportant')

  # Latency Results
  results.Add('average_latency', 'ms',
              Average(s.input_event_latency, s.input_event_count,
                      1000, 3),
              data_type='unimportant')
  results.Add('average_touch_ui_latency', 'ms',
              Average(s.touch_ui_latency, s.touch_ui_count, 1000, 3),
              data_type='unimportant')
  results.Add('average_touch_acked_latency', 'ms',
              Average(s.touch_acked_latency, s.touch_acked_count,
                      1000, 3),
              data_type='unimportant')
  results.Add('average_scroll_update_latency', 'ms',
              Average(s.scroll_update_latency, s.scroll_update_count,
                      1000, 3),
              data_type='unimportant')

File: tools/perf/metrics/smoothness.js
// Copyright 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

'use strict';

/**
 * @fileoverview This file provides the RenderingStats object, used
 * to characterize rendering smoothness.
 */
(function() {
  var getTimeMs = (function() {
    if (window.performance)
      return (performance.now       ||
              performance.mozNow    ||
              performance.msNow     ||
              performance.oNow      ||
              performance.webkitNow).bind(window.performance);
    else
      return function() { return new Date().getTime(); };
  })();

  var requestAnimationFrame = (function() {
    return window.requestAnimationFrame       ||
           window.webkitRequestAnimationFrame ||
           window.mozRequestAnimationFrame    ||
           window.oRequestAnimationFrame      ||
           window.msRequestAnimationFrame     ||
           function(callback) {
             window.setTimeout(callback, 1000 / 60);
           };
  })().bind(window);

  /**
   * Tracks rendering performance using the gpuBenchmarking.renderingStats API.
   * @constructor
   */
  function GpuBenchmarkingRenderingStats() {
  }

  GpuBenchmarkingRenderingStats.prototype.start = function() {
    this.startTime_ = getTimeMs();
    this.initialStats_ = this.getRenderingStats_();
  }

  GpuBenchmarkingRenderingStats.prototype.stop = function() {
    this.stopTime_ = getTimeMs();
    this.finalStats_ = this.getRenderingStats_();
  }

  GpuBenchmarkingRenderingStats.prototype.getStartValues = function() {
    if (!this.initialStats_)
      throw new Error('Start not called.');

    if (!this.finalStats_)
      throw new Error('Stop was not called.');

    return this.initialStats_;
  }

  GpuBenchmarkingRenderingStats.prototype.getEndValues = function() {
    if (!this.initialStats_)
      throw new Error('Start not called.');

    if (!this.finalStats_)
      throw new Error('Stop was not called.');

    return this.finalStats_;
  }

  GpuBenchmarkingRenderingStats.prototype.getDeltas = function() {
    if (!this.initialStats_)
      throw new Error('Start not called.');

    if (!this.finalStats_)
      throw new Error('Stop was not called.');

    var stats = {}
    for (var key in this.finalStats_)
      stats[key] = this.finalStats_[key] - this.initialStats_[key];
    return stats;
  };

  GpuBenchmarkingRenderingStats.prototype.getRenderingStats_ = function() {
    var stats = chrome.gpuBenchmarking.renderingStats();
    stats.totalTimeInSeconds = getTimeMs() / 1000;
    return stats;
  };

  /**
   * Tracks rendering performance using requestAnimationFrame.
   * @constructor
   */
  function RafRenderingStats() {
    this.recording_ = false;
    this.frameTimes_ = [];
  }

  RafRenderingStats.prototype.start = function() {
    if (this.recording_)
      throw new Error('Already started.');
    this.recording_ = true;
    requestAnimationFrame(this.recordFrameTime_.bind(this));
  }

  RafRenderingStats.prototype.stop = function() {
    this.recording_ = false;
  }

  RafRenderingStats.prototype.getStartValues = function() {
    var results = {};
    results.numAnimationFrames = 0;
    results.numFramesSentToScreen = 0;
    results.droppedFrameCount = 0;
    return results;
  }

  RafRenderingStats.prototype.getEndValues = function() {
    var results = {};
    results.numAnimationFrames = this.frameTimes_.length - 1;
    results.numFramesSentToScreen = results.numAnimationFrames;
    results.droppedFrameCount = this.getDroppedFrameCount_(this.frameTimes_);
    return results;
  }

  RafRenderingStats.prototype.getDeltas = function() {
    var endValues = this.getEndValues();
    endValues.totalTimeInSeconds = (
        this.frameTimes_[this.frameTimes_.length - 1] -
        this.frameTimes_[0]) / 1000;
    return endValues;
  };

  RafRenderingStats.prototype.recordFrameTime_ = function(timestamp) {
    if (!this.recording_)
      return;

    this.frameTimes_.push(timestamp);
    requestAnimationFrame(this.recordFrameTime_.bind(this));
  };

  RafRenderingStats.prototype.getDroppedFrameCount_ = function(frameTimes) {
    var droppedFrameCount = 0;
    for (var i = 1; i < frameTimes.length; i++) {
      var frameTime = frameTimes[i] - frameTimes[i-1];
      if (frameTime > 1000 / 55)
        droppedFrameCount++;
    }
    return droppedFrameCount;
  };

  function RenderingStats() {
    if (window.chrome && chrome.gpuBenchmarking &&
        chrome.gpuBenchmarking.renderingStats) {
      return new GpuBenchmarkingRenderingStats();
    }
    return new RafRenderingStats();
  }

  window.__RenderingStats = RenderingStats;
})();
